src/ast/ast_validations.c

```diff
@@ -538,665 +702,748 @@ static AST_Validation _ValidateInlinedProperties(const cypher_astnode_t *props)
-			if(cypher_astnode_type(skip) != CYPHER_AST_INTEGER &&
-			   cypher_astnode_type(skip) != CYPHER_AST_PARAMETER) {
-				ErrorCtx_SetError("SKIP specified value of invalid type, must be a positive integer");
-				return AST_INVALID;
-			}
-		}
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+		return true;
 	}
-	array_free(with_clauses);
 
-	return res;
+	return true;
 }
 
-// A query must end in a RETURN clause, a procedure, or an updating clause
-// (CREATE, MERGE, DELETE, SET, or REMOVE once supported)
-static AST_Validation _ValidateQueryTermination(const AST *ast) {
+// validate a UNION clause
+static bool _Validate_UNION_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		// Make sure all UNIONs specify ALL or none of them do
+		if(vctx->union_all == NOT_DEFINED) {
+			vctx->union_all = cypher_ast_union_has_all(n);
+		}
+		else if(vctx->union_all != cypher_ast_union_has_all(n)) {
+			ErrorCtx_SetError("Invalid combination of UNION and UNION ALL.");
+			vctx->valid = AST_INVALID;
+			return false;
+		}
+
+		// free old bounded vars environment, create a new one
+		vctx->clause = cypher_astnode_type(n);
+		raxFree(vctx->defined_identifiers);
+		vctx->defined_identifiers = raxNew();
+		return true;
+	}
+
+	return true;
+}
+
+// validate a CREATE clause
+static bool _Validate_CREATE_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+		return true;
+	}
+
+	return true;
+}
+
+// validate a MERGE clause
+static bool _Validate_MERGE_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+	}
+
+	return true;
+}
+
+// validate an UNWIND clause
+static bool _Validate_UNWIND_Clause
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID) {
+		return false;
+	}
+
+	if(start) {
+		vctx->clause = cypher_astnode_type(n);
+
+		// introduce alias to bound vars
+		const cypher_astnode_t *alias = cypher_ast_unwind_get_alias(n);
+		const char *identifier = cypher_ast_identifier_get_name(alias);
+		raxInsert(vctx->defined_identifiers, (unsigned char *)identifier, strlen(identifier), NULL, NULL);
+		return true;
+	}
+
+	return true;
```

Comment:
Seems like no validation is performed...
always returns true.

I think we need to discuss pure validations and meta data construction, even though this might cost us a second pass it will give a clear distinction between validations and data gathering